request.go

```diff
@@ -0,0 +1,251 @@
+func (r *MatrixRequest) SetContent(content interface{}) error {
+	if r.fields.Content != nil {
+		return fmt.Errorf("gomatrixserverlib: content already set on the request")
+	}
+	if r.fields.Signatures != nil {
+		return fmt.Errorf("gomatrixserverlib: the request is signed and cannot be modified")
+	}
+	data, err := json.Marshal(content)
+	if err != nil {
+		return err
+	}
+	r.fields.Content = rawJSON(data)
+	return nil
+}
+
+// Method returns the JSON method for the request.
+func (r *MatrixRequest) Method() string {
+	return r.fields.Method
+}
+
+// Content returns the JSON content for the request.
+func (r *MatrixRequest) Content() []byte {
+	return []byte(r.fields.Content)
+}
+
+// Origin returns the server that the request originated on.
+func (r *MatrixRequest) Origin() string {
+	return r.fields.Origin
+}
+
+// RequestURI returns the path and query sections of the HTTP request URL.
+func (r *MatrixRequest) RequestURI() string {
+	return r.fields.RequestURI
+}
+
+// Sign the matrix request with an ed25519 key.
+// Uses the algorithm specified https://matrix.org/docs/spec/server_server/unstable.html#request-authentication
+// Updates the request with the signature in place.
+// Returns an error if there was a problem signing the request.
+func (r *MatrixRequest) Sign(serverName, keyID string, privateKey ed25519.PrivateKey) error {
+	if r.fields.Origin != "" && r.fields.Origin != serverName {
+		return fmt.Errorf("gomatrixserverlib: the request is already signed by a different server")
+	}
+	r.fields.Origin = serverName
+	// The request fields are already in the form required by the specification
+	// So we can just serialise the request fields using the default marshaller
+	data, err := json.Marshal(r.fields)
+	if err != nil {
+		return err
+	}
+	signedData, err := SignJSON(serverName, keyID, privateKey, data)
+	if err != nil {
+		return err
+	}
+	// Now we can deserialise the signed request back into the request structure
+	// to set the Signatures field, (This will clobber the other fields but they
+	// will all round-trip through an encode/decode.)
+	return json.Unmarshal(signedData, &r.fields)
+}
+
+// HTTPRequest constructs an net/http.Request for this matrix request.
+// The request can be passed to net/http.Client.Do().
+func (r *MatrixRequest) HTTPRequest() (*http.Request, error) {
+	urlStr := fmt.Sprintf("matrix://%s%s", r.fields.Destination, r.fields.RequestURI)
+
+	var content io.Reader
+	if r.fields.Content != nil {
+		content = bytes.NewReader([]byte(r.fields.Content))
+	}
+
+	httpReq, err := http.NewRequest(r.fields.Method, urlStr, content)
+	if err != nil {
+		return nil, err
+	}
+
+	if r.fields.Content != nil {
+		httpReq.Header.Set("Content-Type", "application/json")
+	}
+
+	for keyID, sig := range r.fields.Signatures[r.fields.Origin] {
+		httpReq.Header.Add("Authorization", fmt.Sprintf(
+			"X-Matrix origin=\"%s\",key=\"%s\",sig=\"%s\"", r.fields.Origin, keyID, sig,
```

Comment:
No, but only because we rely on the domain and key being safe to include. I guess I should add some checks.